flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -448,14 +441,24 @@ public void failJobDueToTaskFailure(
                         new DispatcherThreadFactory(
                                 Thread.currentThread().getThreadGroup(), "Checkpoint Timer"));
 
+        List<ExecutionVertex> sourceVertices = new ArrayList<>();
```

Comment:
Maybe we could initialize logical to the `CheckpointBriefComputer`, which makes all the related logic together.

Reply:
You could move it to the Computer or put it in a separate method in `ExecutionGraph`, just to make this method here smaller, because it's already quite big.

Reply:
I'll put it in separate method first so that we do not need to modify too much unit tests for now. In the next PR I'll further move it into `CheckpointBriefCalculator`.